lib/webhook.js

```diff
@@ -0,0 +1,84 @@
+const express = require('express');
+const bodyParser = require('body-parser');
+const model = require('./model');
+const initEmitter = require('./emitter');
+
+const handleError = (err, req, res) => {
+  if (err.name === 'SequelizeValidationError') {
+    res.status(400).send(err.message);
+  } else {
+    res.status(500).send(err.message);
+  }
+};
+
+const logger = debug => (req, res, next) => {
+  debug && console.log('Request:', req);
+  next();
+};
+
+const webhook = async (config) => {
+  const subApp = express();
+  const emitter = initEmitter(config);
+  subApp.use(bodyParser.json());
+
+  const router = express.Router();
+  router.use(bodyParser.json());
+  router.use(logger(config.debug));
+
+  router.get('/events', (req, res) => {
+    res.json({
+      events: config.allowedEvents,
+    });
+  });
+
+  router.get('/subscription', async (req, res) => {
+    try {
+      const subscription = await model.getByDiscriminator(req.query[config.discriminatorKey]);
+      res.json(subscription);
+    } catch (err) {
+      handleError(err, req, res);
+    }
+  });
+
+  router.get('/subscription/:id', async (req, res) => {
+    try {
+      const subscription = await model.getById(req.params.id);
+      res.json(subscription);
+    } catch (err) {
+      handleError(err, req, res);
+    }
+  });
+
+  router.post('/subscription', async (req, res) => {
+    try {
+      const subscription = await model.create(req.body);
+      res.json(subscription);
+    } catch (err) {
+      handleError(err, req, res);
+    }
+  });
+
+  router.delete('/subscription/:id', async (req, res) => {
+    try {
+      const subscription = await model.delete(req.params.id);
+      res.json(subscription);
+    } catch (err) {
+      handleError(err, req, res);
+    }
+  });
+
+  router.put('/subscription/:id', async (req, res) => {
+    try {
+      const subscription = await model.update(req.params.id, req.body);
+      res.json(subscription);
+    } catch (err) {
+      handleError(err, req, res);
+    }
+  });
+
+  await model.init(config);
```

Comment:
You could make modelPromise and store that. Then in each route handler functions you can do, modelPromise.then(model.create(req.body) etc. 
That way, the web hook function will no longer need to be async and we would be able to use it in a much more natural manner,
app.use('/api/v1', webhook.router);
Other handlers may need the emitter to fire events. What could be a good way for them to access it?

Reply:
Yes, I agree this looks more natural way to use. Earlier I had done it similar to how you mentioned above. I only thing was if the model init fails then there is no way to notify and the api calls on these routes would be failing, unless we pass another callback function to notify back status of model init.

Reply:
We can add a reject handler to the model promise to handle the model init error.

Reply:
My question was more on, should we propagate this error back to consumer?

Reply:
Just log it. It should be treated as misconfiguration. Even if we were to invoke an application specified callback, there is nothing that the application developer can do in the callback, to fix the issue at run time.